Userland/Libraries/LibGUI/TextDocument.cpp

```diff
@@ -383,6 +407,96 @@ DeprecatedString TextDocument::text_in_range(TextRange const& a_range) const
     return builder.to_deprecated_string();
 }
 
+// This function will determine the number of codepoints of the glyph immediately preceding the cursor, by determining if there is an emoji. If no emoji is found it will return 1.
+size_t TextDocument::get_code_points_before_cursor(TextPosition const& cursor) const
+{
+    if (!cursor.is_valid())
+        return 0;
+
+    return get_prev_grapheme_cluster_boundary(cursor);
+}
+
+// This function will return the position of the next grapheme cluster break, relative to the cursor, for "correct looking" parsing of unicode based on grapheme cluster boundary algorithm.
+size_t TextDocument::get_next_grapheme_cluster_boundary(TextPosition const& cursor) const
+{
+    if (!cursor.is_valid())
+        return 0;
+
+    auto& line = this->line(cursor.line());
+    auto line_span = Span<u32 const>(line.code_points(), line.length());
+
+    if (line_span.is_empty())
+        return 0;
+
+    if (cursor.column() == line.length())
+        return 0;
+
+    auto boundaries_cache = Unicode::find_grapheme_segmentation_boundaries(Utf32View(line.code_points(), line.length()));
+
+    if (boundaries_cache.is_empty())
+        return cursor.column();
+
+    if (boundaries_cache.size() == 1)
+        return cursor.column() - boundaries_cache.at(0);
+
+    auto previous_boundary_guess = boundaries_cache.at(0);
+    auto boundary_guess = boundaries_cache.at(1);
+
+    for (auto current_boundary_guess : boundaries_cache) {
+        // If the current boundary is further to the left, update our guess
+        if (current_boundary_guess < boundary_guess) {
+            previous_boundary_guess = boundary_guess;
+            boundary_guess = current_boundary_guess;
+        }
+    }
+
+    // Assume that the cursor cannot sit at a point that is not a boundary, so return previous guess.
+    return previous_boundary_guess;
```

Comment:
This doesn't seem right. `find_grapheme_segmentation_boundaries` returns indices within the view, which is currently the entire line the cursor is on. This loop doesn't consult the index of the cursor at all, so I don't see how it can return "the position of the next grapheme cluster break, relative to the cursor".

The indices returned by `find_grapheme_segmentation_boundaries` will also always be in order. So this will always return `boundaries_cache[1]`, because `current_boundary_guess < boundary_guess` won't be true after it hits index 1. This will also have weird behavior at the beginning of the loop because you are looking at indices 0 and 1 again.

Reply:
Okay I addressed this by taking the logic from `get_previous_grapheme_cluster_boundary` (previously `get_code_points_before_cursor`) and reversed `boundaries_cache`. This seems to work nicely and clean up some of the quirks of the way it was done previously